### PR TITLE
Switch from boost::MPL to brigand and revamp travis ci

### PR DESCRIPTION
**Pull request by jbakosi**
This takes care of #83, now we can also do #4. Thanks @junghans for the help with travis and docker. Please add back the push.

---
This change is [Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/254)

### CODE REVIEW

**Comment by jbakosi** [comment]
Alright @junghans , I think this is in good shape now.

File: .travis.yml [hunk 1]
```diff
@@ -6,60 +6,52 @@ services:
...
+                 -f ${HOME}/docker/Dockerfile.quinoa-build-travis
+                 ${HOME}/docker/ &&
+    CON=$(docker run -d ${TRAVIS_REPO_SLUG}:${DISTRO} /bin/bash) &&
     docker cp ${CON}:/home/quinoa/.sonar ${HOME}/
 
-after_success:
-  - if [[ ${DOCKERHUB} = true && ${DOCKER_USERNAME} && ${DOCKER_PASSWORD} && ${TRAVIS_PULL_REQUEST} == false && ${TRAVIS_BRANCH} == master ]]; then
-      docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD";
-      docker push "${TRAVIS_REPO_SLUG}:${DISTRO}${TAG}";
-    fi
```
Comment (junghans) [inline .travis.yml]: Didn't you want to put that back?

File: .travis.yml [hunk 2]
```diff
@@ -6,60 +6,52 @@ services:
...
+                 --build-arg TRAVIS_JOB_NUMBER=${TRAVIS_JOB_NUMBER}
+                 --build-arg TRAVIS_PULL_REQUEST=${TRAVIS_PULL_REQUEST}
+                 --build-arg TRAVIS_JOB_ID=${TRAVIS_JOB_ID}
+                 --build-arg TRAVIS_REPO_SLUG=${TRAVIS_REPO_SLUG}
+                 --build-arg TRAVIS_COMMIT=${TRAVIS_COMMIT}
+                 -t ${TRAVIS_REPO_SLUG}:${DISTRO}
+                 -f ${HOME}/docker/Dockerfile.quinoa-build-travis
+                 ${HOME}/docker/ &&
+    CON=$(docker run -d ${TRAVIS_REPO_SLUG}:${DISTRO} /bin/bash) &&
     docker cp ${CON}:/home/quinoa/.sonar ${HOME}/
```
Comment (junghans) [inline .travis.yml]: It would be worth testing how much that saves!

File: docker/Dockerfile.quinoa-build-travis [hunk 3]
```diff
@@ -18,10 +16,13 @@ ARG TRAVIS_BRANCH
 ARG TRAVIS_JOB_NUMBER
 ARG TRAVIS_PULL_REQUEST
 ARG TRAVIS_JOB_ID
-ARG TRAVIS_TAG
 ARG TRAVIS_REPO_SLUG
 ARG TRAVIS_COMMIT
 
+ENV RUNNER_ARGS "--bind-to none -oversubscribe"
+ENV OMPI_MCA_plm isolated
+ENV PATH /opt/openmpi/${COMPILER}/bin:$PATH
```
Comment (junghans) [inline docker/Dockerfile.quinoa-build-travis]: Nice!

File: .travis.yml (hunk 1 above)
Comment (jbakosi) [reply .travis.yml]: Yeah! I think so.

Comment (jbakosi) [reply .travis.yml]: Can you just push it?

File: .travis.yml (hunk 2 above)
Comment (jbakosi) [reply .travis.yml]: Indeed.

File: docker/Dockerfile.quinoa-build-travis (hunk 3 above)
Comment (jbakosi) [reply docker/Dockerfile.quinoa-build-travis]: Yeah, since all builds now run inside the same image.

File: .travis.yml (hunk 1 above)
Comment (jbakosi) [reply .travis.yml]: Yeah. I misunderstood your question. I will out this back and merge. Thanks for taking a look.

**Pull request merged**